modules/biobank/php/specimendao.class.inc

```diff
@@ -0,0 +1,719 @@
+     * query the referenced tables and columns and build a nested array of
+     * reference table option values, with the Reference Table ID as the index.
+     *
+     * @return array $attributeOptions A nested array of attribute reference
+     *                                 table options indexed by Reference Table
+     *                                 ID
+     */
+    public function getAttributeOptions() : array
+    {
+        $query = 'SELECT ReferenceTableID as id,
+                         TableName as tableName,
+                         ColumnName as columnName
+                  FROM biobank_reference_table';
+        $attributeOptionsReferences = $this->db->pselect($query, array());
+
+        $attributeOptions = array();
+        foreach ($attributeOptionsReferences as $reference) {
+            $query  = 'SELECT '.$reference['columnName'].' FROM '.$reference['tableName'];
+            $result = $this->db->pselect($query, array());
+
+            $options = array();
+            foreach ($result as $row) {
+                $options[$row[$reference['columnName']]] = $row[$reference['columnName']];
+            }
+
+            $attributeOptions[$reference['id']] = $options;
+        }
+
+        return $attributeOptions;
+    }
+
+
+    // XXX: This should potentially be queried from a table called
+    // `biobank_specimen_units` since containers are not currently using units.
+    // Discussions need to be had with Rida regarding this.
+    /**
+     * Queries all rows of the biobank_unit table and returns a nested
+     * array with the ID field as the index.
+     *
+     * @return array $specimenUnits All data concerning each unit
+     */
+    public function getUnits() : array
+    {
+        $query = "SELECT UnitID as id,
+                         Label as unit 
+                  FROM biobank_unit";
+        $units = $this->db->pselectWithIndexKey($query, array(), 'id');
+
+        return $units;
+    }
+
+    /**
+     * Queries all rows of the biobank_specimen_type_unit_rel table and returns
+     * a nested array of specimen unit values, with the Type ID as the first index,
+     * and the Unit ID as the second index.
+     *
+     * @return array $specimenTypeUnits A nested array of unit values index by
+     *                                  Type ID and Unit ID
+     */
+    public function getTypeUnits() : array
+    {
+        $query     = "SELECT bst.SpecimenTypeID as typeId,
+                          bu.UnitID as unitId,
+                          bu.Label as unit
+                  FROM biobank_specimen_type_unit_rel bstu
+                  JOIN biobank_specimen_type bst USING (SpecimenTypeID)
+                  JOIN biobank_unit bu USING (UnitID)";
+        $result    = $this->db->pselect($query, array());
+        $typeUnits = array();
+        foreach ($result as $row) {
+            foreach ($row as $value) {
+                $typeUnits[$row['typeId']][$row['unitId']]['unit'] = $row['unit'];
```

Comment:
Yeah, I think the idea is that if this table were to expand to have more properties (e.g. `system`: 'metric'), then it would be useful to maintain this structure.

Reply:
I see